multimap/src/main/java/org/infinispan/multimap/impl/ListBucket.java

```diff
@@ -220,6 +220,36 @@ public Collection<Long> indexOf(V element, long count, long rank, long maxLen) {
       return positions;
    }
 
+   public ListBucket<V> insert(boolean before, V pivot, V element) {
+      Deque<V> newValues = new ArrayDeque<>(values.size() +1);
+
+      Iterator<V> iterator = values.iterator();
+      boolean found = false;
+      while (iterator.hasNext()) {
+         V next = iterator.next();
+         if (found) {
+            newValues.offerLast(next);
+         } else {
+            if (Objects.deepEquals(pivot, next))  {
+               found = true;
+               if (before) {
+                  newValues.offerLast(element);
+                  newValues.offerLast(next);
+               } else {
+                  newValues.offerLast(next);
+                  newValues.offerLast(element);
+               }
+            } else {
+               newValues.offerLast(next);
+            }
+         }
+      }
+      if (!found) {
+         return new ListBucket<>();
```

Comment:
Returning `null` instead would avoid allocating both the `ListBucket` and an `ArrayDeque`, and also would be simpler to check if the operation was applied on the caller side.